sdk/python/kfp/components/_dsl_bridge.py

```diff
@@ -13,160 +13,70 @@
-                return str(input_value)
-            else:
-                input_spec = inputs_dict[input_name]
-                if input_spec.optional:
-                    return None
-                else:
-                    raise ValueError('No value provided for input {}'.format(input_name))
-
-        if isinstance(arg, InputPathPlaceholder):
-            input_name = arg.input_name
-            input_value = argument_values.get(input_name, None)
-            if input_value is not None:
-                input_path = _generate_input_file_name(input_name)
-                input_paths[input_name] = input_path
-                artifact_arguments[input_name] = input_value
-                return input_path
-            else:
-                input_spec = inputs_dict[input_name]
-                if input_spec.optional:
-                    #Even when we support default values there is no need to check for a default here.
-                    #In current execution flow (called by python task factory), the missing argument would be replaced with the default value by python itself.
-                    return None
-                else:
-                    raise ValueError('No value provided for input {}'.format(input_name))
-
-        elif isinstance(arg, OutputPathPlaceholder):
-            output_name = arg.output_name
-            output_filename = _generate_output_file_name(output_name)
-            if arg.output_name in output_paths:
-                if output_paths[output_name] != output_filename:
-                    raise ValueError('Conflicting output files specified for port {}: {} and {}'.format(output_name, output_paths[output_name], output_filename))
-            else:
-                output_paths[output_name] = output_filename
-
-            return output_filename
-
-        elif isinstance(arg, ConcatPlaceholder):
-            expanded_argument_strings = expand_argument_list(arg.items)
-            return ''.join(expanded_argument_strings)
-
-        elif isinstance(arg, IfPlaceholder):
-            arg = arg.if_structure
-            condition_result = expand_command_part(arg.condition)
-            from distutils.util import strtobool
-            condition_result_bool = condition_result and strtobool(condition_result) #Python gotcha: bool('False') == True; Need to use strtobool; Also need to handle None and []
-            result_node = arg.then_value if condition_result_bool else arg.else_value
-            if result_node is None:
-                return []
-            if isinstance(result_node, list):
-                expanded_result = expand_argument_list(result_node)
-            else:
-                expanded_result = expand_command_part(result_node)
-            return expanded_result
+def _create_container_op_from_component_and_arguments(
+    component_spec: ComponentSpec,
+    arguments: Mapping[str, Any],
+    component_ref: ComponentReference = None,
+): # -> 'ContainerOp':
```

Comment:
Why can't we do type hint here?

Reply:
Making a proper type hint requires importing from `kfp.dsl` on the module level.
Doing so will create a circular reference since `kfp.dsl` already depends on `kfp.components`.

The way to fix this in the future is to move `kfp/components/_dsl_bridge.py` to `kfp/dsl/`. I did not do that in this PR to reduce the refactoring diff size.

Reply:
Yes it is indeed a problem... Maybe we can do forward reference here for now?

Reply:
> Maybe we can do forward reference here for now?

Forward reference still requires an import - otherwise flake8 complains of undefined identifier. See https://travis-ci.com/kubeflow/pipelines/jobs/276973031

If you want I can restore the forward reference and disable flake check for this line.

Reply:
Managed to fix this.

Reply:
Thanks!